### PR TITLE
Gotestsum

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Switch to https://github.com/gotestyourself/gotestsum for an improved testing experience:

CleanShot 2023-06-08 at 21 04 25@2x

CleanShot 2023-06-08 at 21 05 52@2x

##### Highlights

* use gotestsum (itself using the built-in test2json) to programmatically parse the Go output and generate something far less verbose for CI consumption
* keep the fully verbose output available via the `gotestsum.json` artifact
* generate working JUnit XML summaries and integrate with GitHub actions to surface build errors without having to look at the logs
* less magic! more fun!

#### Ticket Link
None.

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/Makefile [hunk 1]
```diff
@@ -54,10 +54,17 @@ ifeq ($(BUILD_NUMBER),dev)
 	export MM_FEATUREFLAGS_GRAPHQL = true
 endif
 
+# Ensure developer invocation and tests are anchored.
+MM_SERVER_PATH ?= $(ROOT)
+
+# Go test sum configuration
+GOTESTSUM_FORMAT ?= testname
+GOTESTSUM_JUNITFILE ?= report.xml
+GOTESTSUM_JSONFILE ?= gotestsum.json
```
Comment (lieut-data) [inline server/Makefile:63]: Some opinionated defaults for using `gotestsum.`

File: server/Makefile [hunk 2]
```diff
@@ -54,10 +54,17 @@ ifeq ($(BUILD_NUMBER),dev)
...
+
 # mmctl
-MM_SERVER_PATH ?= $(PWD)
 MMCTL_BUILD_TAGS =
-MMCTL_TESTFLAGS = -timeout 30m -race -v
+MMCTL_TESTFLAGS = -timeout 30m -race
```
Comment (lieut-data) [inline server/Makefile:67]: `-v` is handled by `gotestsum` now.

File: server/Makefile [hunk 3]
```diff
@@ -457,17 +458,28 @@ modules-tidy:
...
 ifeq ($(BUILD_ENTERPRISE_READY),true)
 	@echo Running all tests
 else
 	@echo Running only TE tests
 endif
 
+test-server-race: export MM_SERVER_PATH := $(MM_SERVER_PATH)
+test-server-race: export GOTESTSUM_FORMAT := $(GOTESTSUM_FORMAT)
+test-server-race: export GOTESTSUM_JUNITFILE := $(GOTESTSUM_JUNITFILE)
+test-server-race: export GOTESTSUM_JSONFILE := $(GOTESTSUM_JSONFILE)
```
Comment (lieut-data) [inline server/Makefile:471]: This is perhaps the ugliest part of these changes, as I have to duplicate them for every test. I opted not to clutter the command invocations themselves, and I discovered I can't do this in just `test-server-pre` as it only applies to the rule. There's other tricks like exporting all `Makefile` vars into the envrionment all the time, but I didn't want to risk unexpected breakages here.

Honestly, longer term, I'm wondering if we consider something like Mage or Taskfile to do this more cleanly anyway.

File: server/Makefile [hunk 4]
```diff
@@ -457,17 +458,28 @@ modules-tidy:
...
 test-server-race: test-server-pre
-	./scripts/test.sh "$(GO)" "-race $(GOFLAGS)" "$(TE_PACKAGES) $(EE_PACKAGES)" "$(TESTS)" "$(TESTFLAGS)" "$(GOBIN)" "90m" "atomic"
-	./scripts/test.sh "$(GO)" "-race $(GOFLAGS)" "$(BOARDS_PACKAGES)" "$(TESTS)" "$(TESTFLAGS)" "$(GOBIN)" "90m" "atomic"
-	./scripts/test.sh "$(GO)" "-race $(GOFLAGS)" "$(PLAYBOOKS_PACKAGES)" "$(TESTS)" "$(TESTFLAGS)" "$(GOBIN)" "90m" "atomic"
+ifeq ($(IS_CI),true)
```
Comment (lieut-data) [inline server/Makefile:473]: Do the magic re: `IS_CI` here vs. in the GitHub actions.

**Comment by lieut-data** [comment]
Note that the log output isn't 100% clean, as we still get a bunch of "morph spam" from logs that get emitted outside of a test context. This is something I'll chip away at later to streamline things even more, but the big rocks are done.

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:   Great improvement.

**Review by crspeller — APPROVED** [review]
Nice!

**Pull request merged**